File: scripts/src/bin/writing.rs
use color_eyre::{eyre::eyre, Result};
use dialoguer::{theme::ColorfulTheme, Confirm, Input, MultiSelect, Select};
use scripts::{get_current_date, get_file_info};
use serde::Serialize;
use slug::slugify;
use std::fs;
use titlecase::titlecase;

const TYPE_CHOICES: [&str; 3] = ["essay", "tutorial", "note"];
const TAGS_CHOICES: [&str; 7] = [
    "Web Development",
    "React",
    "Developer Tools",
    "Programming",
    "Self-Hosting",
    "General",
    "Outdoors",
];
const ICON_CHOICES: [&str; 9] = [
    "web-development",
    "react",
    "developer-tools",
    "programming",
    "self-hosting",
    "general",
    "ai",
    "outdoors",
    "travel",
];

const BODY_TEMPLATE: &str = r#"

Context: What are you talking about?
Relevance: Why should I care about that?
Application: How do I do it?

Learning goals => Assessment => Lesson Plan
"#;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct Frontmatter {
    title: String,
    slug: String,
    description: String,
    date: String,
    last_updated: String,
    r#type: String,
    tags: Vec<String>,
    icon: String,
}

fn main() -> Result<()> {
    color_eyre::install()?;

    let theme = ColorfulTheme::default();
    let current_date = get_current_date();

    let title: String = Input::with_theme(&theme)
        .with_prompt("Title")
        .interact_text()?;

    let slug: String = Input::with_theme(&theme)
        .with_prompt("Slug")
        .default(slugify(&title))
        .interact_text()?;

    let description: String = Input::with_theme(&theme)
        .with_prompt("Description")
        .interact_text()?;

    let date: String = Input::with_theme(&theme)
        .with_prompt("Date")
        .default(current_date.to_string())
        .interact_text()?;

    let chosen_type = Select::with_theme(&theme)
        .with_prompt("Pick a type")
        .items(&TYPE_CHOICES)
        .interact()
        .map(|choice| TYPE_CHOICES[choice])?;

    let tags = MultiSelect::with_theme(&theme)
        .with_prompt("Choose your tags")
        .items(&TAGS_CHOICES)
        .interact()?
        .into_iter()
        .map(|tag_index| TAGS_CHOICES[tag_index]);

    let icon = Select::with_theme(&theme)
        .with_prompt("Pick an icon")
        .items(&ICON_CHOICES)
        .interact()
        .map(|choice| ICON_CHOICES[choice])?;

    let file_info = get_file_info(current_date.clone(), slug.clone())?;

    let fm = Frontmatter {
        title: titlecase(&title),
        slug,
        date,
        last_updated: current_date.clone(),
        description,
        r#type: chosen_type.to_string(),
        icon: icon.to_string(),
        tags: tags.map(|s| s.to_string()).collect::<Vec<String>>(),
    };

    let frontmatter = format!("---\n{}---", serde_yaml::to_string(&fm)?);

    println!(
        r#"
The file "{}" will be created with:

{}
    "#,
        file_info.filename, frontmatter
    );

    if !Confirm::with_theme(&theme)
        .with_prompt("Do you want to continue?")
        .interact()?
    {
        return Err(eyre!("Aborting..."));
    }

    let file_content = format!("{}{}", frontmatter, BODY_TEMPLATE);

    fs::write(&file_info.filepath, file_content)?;

    println!("Successfully created {}", file_info.filepath.display());

    Ok(())
}

File: scripts/writing.ts
import fs from 'node:fs/promises'
import path from 'node:path'
import { confirm, group, intro, isCancel, multiselect, outro, select, text } from '@clack/prompts'
import slugify from '@sindresorhus/slugify'
import { ICON_CHOICES, TAGS_CHOICES, TYPE_CHOICES } from '../src/constants/content'
import { titleCase } from './titlecase'

type Icon = typeof ICON_CHOICES[number]
type Tag = typeof TAGS_CHOICES[number]
type Type = typeof TYPE_CHOICES[number]

const BODY_TEMPLATE = `Context: What are you talking about?
Relevance: Why should I care about that?
Application: How do I do it?

Learning goals => Assessment => Lesson Plan`

interface Frontmatter {
	title: string
	slug: string
	description: string
	date: string
	lastUpdated: string
	type: Type
	tags: Tag[]
	icon: Icon
}

async function main() {
	intro('Writing Wizard')

	const rawTitle = await text({
		message: 'Title',
		validate(value) {
			if (!value)
				return 'Title is required'
		},
	})

	if (isCancel(rawTitle)) {
		outro('Bye!')
		process.exit(0)
	}

	const rest = await group({
		slug: () => text({
			message: 'Slug',
			initialValue: slugify(rawTitle, { separator: '-' }),
			validate(value) {
				if (!value)
					return 'Slug is required'
			},
		}),
		description: () => text({
			message: 'Description',
			validate(value) {
				if (!value)
					return 'Description is required'
			},
		}),
		date: () => text({
			message: 'Date',
			initialValue: new Date().toISOString().split('T')[0],
			validate(value) {
				if (!value)
					return 'Date is required'

				if (!/^\d{4}-\d{2}-\d{2}$/.test(value))
					return 'Invalid date format'
			},
		}),
		type: () => select({
			message: 'Pick a type',
			options: TYPE_CHOICES.map(type => ({ value: type, label: type })),
		}),
		tags: () => multiselect({
			message: 'Pick tags',
			options: TAGS_CHOICES.map(tag => ({ value: tag, label: tag })),
			required: true,
		}),
		icon: () => select({
			message: 'Pick an icon',
			options: ICON_CHOICES.map(icon => ({ value: icon, label: icon })),
		}),
	})

	if (isCancel(rest)) {
		outro('Bye!')
		process.exit(0)
	}

	const fm: Frontmatter = {
		title: titleCase(rawTitle),
		lastUpdated: rest.date,
		...rest,
	}

	const filename = `${fm.date}--${fm.slug}.mdx`
	const filePath = path.join(process.cwd(), 'src', 'content', 'writing', filename)

	const content = `---
title: "${fm.title}"
slug: ${fm.slug}
description: "${fm.description}"
date: ${fm.date}
lastUpdated: ${fm.lastUpdated}
type: ${fm.type}
icon: ${fm.icon}
tags:
${fm.tags.map(tag => `  - ${tag}`).join('\n')}
---

${BODY_TEMPLATE}
`

	const shouldProceed = await confirm({
		message: 'Want to create the file?',
	})

	if (!shouldProceed) {
		outro('Bye!')
		process.exit(0)
	}

	await fs.writeFile(filePath, content)

	outro(`File created at ${filePath}`)
}

main().catch((error) => {
	console.error('Error:', error)
	process.exit(1)
})
